src/raft.c

```diff
@@ -164,6 +354,131 @@ void RaftExecuteCommandArray(RedisModuleCtx *ctx,
+    size_t num_keys;
+    RedisModuleString ** keys = RaftRedisLockKeysDeserialize(entry->data, entry->data_len, &num_keys);
+
+    /* sanity check keys all belong to same slot */
+    int slot = -1;
+    for (size_t i = 0; i < num_keys; i++) {
+        RedisModuleString * key = keys[i];
+
+        unsigned int thisslot = keyHashSlotRedisString(key);
+        if (slot == -1) {
+            slot = (int) thisslot;
+        } else {
+            if (slot != (int) thisslot) {
+                if (req) {
+                    replyCrossSlot(req->ctx);
+                }
+                goto error;
+            }
+        }
+    }
+
+    if (slot == -1) { /* should be impossible, as keys should be listed up front */
+        if (req) {
+            RedisModule_ReplyWithSimpleString(req->ctx, "OK");
+        }
+        goto error;
+    }
+
+    ShardingInfo *si = rr->sharding_info;
+    if (!si->migrating_slots_map[slot]) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR keys are not migratable");
+        }
+        goto error;
+    }
+
+    if (!si->migrating_slots_map[slot]->local) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR This RedisRaft cluster doesn't own these keys");
+        }
+        goto error;
+    }
+
+    if (!si->importing_slots_map[slot]) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR no RedisRaft cluster to import keys into");
+        }
+        goto error;
+    }
+
+    if (si->importing_slots_map[slot]->local) {
+        if (req) {
+            RedisModule_ReplyWithError(req->ctx, "ERR trying to import keys into self RedisCluster");
+        }
+        goto error;
+    }
+
+    for (size_t i = 0; i < num_keys; i++) {
+        if (RedisModule_KeyExists(rr->ctx, keys[i])) {
+/*            size_t str_len;
+            const char *str = RedisModule_StringPtrLen(keys[i], &str_len);
+            LOG_WARNING("locking %.*s", (int) str_len, str);
+*/
+            RedisModule_DictSet(rr->locked_keys, keys[i], NULL);
+        }
+    }
+
+    if (req) {
+        memcpy(req->r.migrate_keys.shardGroupId, si->importing_slots_map[slot]->id, RAFT_DBID_LEN);
+        MigrateKeys(rr, req);
+    }
+    goto exit;
+
+error:
+    if (req) {
+        RaftReqFree(req);
+    }
+
+exit:
+    for (size_t i = 0; i < num_keys; i++) {
+        RedisModule_FreeString(rr->ctx, keys[i]);
+    }
+    RedisModule_Free(keys);
+}
+
+static void unlockDeleteKeys(RedisRaftCtx *rr, raft_entry_t *entry)
+{
+    RedisModule_Assert(entry->type == RAFT_LOGTYPE_DELETE_UNLOCK_KEYS);
+
+    RaftReq *req = entry->user_data;
+
+    size_t num_keys;
+    RedisModuleString ** keys;
+
+    keys = RaftRedisLockKeysDeserialize(entry->data, entry->data_len, &num_keys);
+
+    enterRedisModuleCall();
+    RedisModuleCallReply *reply;
+    RedisModule_Assert((reply = RedisModule_Call(redis_raft.ctx, "del", "v", keys, num_keys)) != NULL);
```

Comment:
```suggestion
    reply = RedisModule_Call(redis_raft.ctx, "del", "v", keys, num_keys);
    RedisModule_Assert(reply);
```

RedisModule_Assert() is a runtime check but traditionally, `assert()` is not part of release builds. So, having a function in the assert suggests that it's not a critical function, in other words, it's just there for testing. So, I suggest moving it out.